File: apps/backend/kubedeck_backend/api/watch_manager.py
from __future__ import annotations

import os
import subprocess
import threading
import time
import uuid
from collections import deque
from dataclasses import dataclass, field
from typing import Any, Callable, Deque

from kubedeck_backend.kubectl.command import KubectlCommand, kubectl_environment
from kubedeck_backend.api.resource_cache import apply_watch_event_line_to_resource_cache
from kubedeck_backend.api.watch_events import watch_event_hub

WATCH_OUTPUT_TAIL_LINES = 20
WATCH_STOP_TIMEOUT_SECONDS = 3.0

PopenFactory = Callable[..., subprocess.Popen[str]]


def normalize_watch_namespace(namespace: str | None) -> str:
    text = str(namespace or "all").strip()
    return text or "all"


def build_watch_args(resource: str, namespace: str) -> list[str]:
    args = ["get", resource, "-o", "json", "--watch=true", "--output-watch-events=true"]
    if namespace == "all":
        args.append("-A")
    elif namespace != "_cluster":
        args.extend(["-n", namespace])
    return args


@dataclass(frozen=True)
class WatchKey:
    cluster_id: str
    resource: str
    namespace: str

    def label(self) -> str:
        return f"{self.cluster_id}:{self.namespace}:{self.resource}"


@dataclass
class WatchSession:
    id: str
    key: WatchKey
    command_preview: str
    process: Any
    started_at: float = field(default_factory=time.time)
    updated_at: float = field(default_factory=time.time)
    status: str = "running"
    stdout_lines: int = 0
    stderr_lines: int = 0
    cache_events: int = 0
    cache_invalidations: int = 0
    exit_code: int | None = None
    stopped_by_user: bool = False
    output_tail: Deque[str] = field(default_factory=lambda: deque(maxlen=WATCH_OUTPUT_TAIL_LINES))
    error_tail: Deque[str] = field(default_factory=lambda: deque(maxlen=WATCH_OUTPUT_TAIL_LINES))
    _lock: threading.RLock = field(default_factory=threading.RLock)

    def remember_stdout(self, line: str) -> None:
        with self._lock:
            self.stdout_lines += 1
            self.updated_at = time.time()
            text = (line or "").rstrip()
            if text:
                self.output_tail.append(text[:1000])

    def remember_stderr(self, line: str) -> None:
        with self._lock:
            self.stderr_lines += 1
            self.updated_at = time.time()
            text = (line or "").rstrip()
            if text:
                self.error_tail.append(text[:1000])

    def remember_cache_result(self, result: dict[str, Any]) -> None:
        if not result.get("parsed"):
            return
        with self._lock:
            self.cache_events += 1
            self.cache_invalidations += int(result.get("cleared") or 0)
            self.updated_at = time.time()

    def sync_status(self) -> None:
        poll = getattr(self.process, "poll", None)
        if not callable(poll):
            return
        code = poll()
        if code is None:
            return
        with self._lock:
            self.exit_code = code
            if self.status == "running":
                self.status = "stopped" if self.stopped_by_user or code == 0 else "failed"
            self.updated_at = time.time()

    def view(self) -> dict[str, Any]:
        self.sync_status()
        with self._lock:
            return {
                "id": self.id,
                "clusterId": self.key.cluster_id,
                "resource": self.key.resource,
                "namespace": self.key.namespace,
                "status": self.status,
                "pid": getattr(self.process, "pid", None),
                "startedAt": self.started_at,
                "updatedAt": self.updated_at,
                "ageSeconds": max(0.0, time.time() - self.started_at),
                "stdoutLines": self.stdout_lines,
                "stderrLines": self.stderr_lines,
                "cacheEvents": self.cache_events,
                "cacheInvalidations": self.cache_invalidations,
                "exitCode": self.exit_code,
                "stoppedByUser": self.stopped_by_user,
                "commandPreview": self.command_preview,
                "outputTail": list(self.output_tail),
                "errorTail": list(self.error_tail),
            }


class ResourceWatchManager:
    """Lifecycle foundation for Kubernetes watch -> cache -> WebSocket updates.

    Watch output is connected to ResourceSnapshotCache invalidation and parsed
    events are now published to the WebSocket event hub. HTTP polling remains
    the safe frontend fallback.
    """

    def __init__(self, popen_factory: PopenFactory | None = None) -> None:
        self._popen_factory = popen_factory or subprocess.Popen
        self._sessions_by_id: dict[str, WatchSession] = {}
        self._sessions_by_key: dict[WatchKey, str] = {}
        self._lock = threading.RLock()

    def start(self, command: KubectlCommand, resource: str, namespace: str) -> dict[str, Any]:
        key = WatchKey(cluster_id=command.cluster_id, resource=resource, namespace=normalize_watch_namespace(namespace))
        with self._lock:
            existing_id = self._sessions_by_key.get(key)
            if existing_id:
                existing = self._sessions_by_id.get(existing_id)
                if existing:
                    existing.sync_status()
                    if existing.status == "running":
                        view = existing.view()
                        view["alreadyRunning"] = True
                        return view

            process = self._spawn(command)
            session = WatchSession(
                id=str(uuid.uuid4()),
                key=key,
                command_preview=command.preview,
                process=process,
            )
            self._sessions_by_id[session.id] = session
            self._sessions_by_key[key] = session.id
            self._start_reader_threads(session)
            view = session.view()
            view["alreadyRunning"] = False
            return view

    def stop(self, watch_id: str) -> dict[str, Any]:
        with self._lock:
            session = self._sessions_by_id.get(watch_id)
        if session is None:
            return {"ok": False, "found": False, "id": watch_id}
        self._stop_session(session)
        return {"ok": True, "found": True, "watch": session.view()}

    def stop_all(self) -> dict[str, Any]:
        with self._lock:
            sessions = list(self._sessions_by_id.values())
        stopped = 0
        for session in sessions:
            if session.view()["status"] == "running":
                self._stop_session(session)
                stopped += 1
        return {"ok": True, "stopped": stopped, "watches": [session.view() for session in sessions]}

    def status(self) -> dict[str, Any]:
        with self._lock:
            sessions = list(self._sessions_by_id.values())
        views = [session.view() for session in sessions]
        running = sum(1 for item in views if item.get("status") == "running")
        return {
            "enabled": True,
            "mode": "cache-invalidation+websocket-events",
            "running": running,
            "total": len(views),
            "watches": views,
            "note": "Running watches parse Kubernetes watch events, invalidate affected resource-list cache entries, and publish lightweight WebSocket events. HTTP polling remains the fallback.",
        }

    def _spawn(self, command: KubectlCommand) -> Any:
        creationflags = 0
        if os.name == "nt" and hasattr(subprocess, "CREATE_NO_WINDOW"):
            creationflags = subprocess.CREATE_NO_WINDOW
        return self._popen_factory(
            command.argv,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            stdin=subprocess.DEVNULL,
            text=True,
            encoding="utf-8",
            errors="replace",
            shell=False,
            env=kubectl_environment(command.kubeconfig_path),
            creationflags=creationflags,
        )

    def _start_reader_threads(self, session: WatchSession) -> None:
        stdout = getattr(session.process, "stdout", None)
        stderr = getattr(session.process, "stderr", None)
        if stdout is not None:
            thread = threading.Thread(target=self._read_stream, args=(session, stdout, False), daemon=True)
            thread.start()
        if stderr is not None:
            thread = threading.Thread(target=self._read_stream, args=(session, stderr, True), daemon=True)
            thread.start()

    def _read_stream(self, session: WatchSession, stream: Any, is_stderr: bool) -> None:
        try:
            for line in iter(stream.readline, ""):
                if not line:
                    break
                if is_stderr:
                    session.remember_stderr(line)
                else:
                    session.remember_stdout(line)
                    try:
                        result = apply_watch_event_line_to_resource_cache(
                            session.key.cluster_id,
                            session.key.resource,
                            session.key.namespace,
                            line,
                        )
                        session.remember_cache_result(result)
                        if result.get("parsed"):
                            watch_event_hub.publish({
                                "type": "resource.changed",
                                "clusterId": session.key.cluster_id,
                                "watchId": session.id,
                                "resource": result.get("resource") or session.key.resource,
                                "namespace": result.get("namespace") or session.key.namespace,
                                "name": result.get("name") or "",
                                "eventType": result.get("eventType") or "OBJECT",
                                "cacheInvalidations": int(result.get("cleared") or 0),
                            })
                    except Exception as exc:  # pragma: no cover - defensive cache integration guard
                        session.remember_stderr(f"watch cache integration failed: {exc}")
        except Exception as exc:  # pragma: no cover - defensive thread guard
            session.remember_stderr(f"watch reader failed: {exc}")
        finally:
            session.sync_status()

    def _stop_session(self, session: WatchSession) -> None:
        with session._lock:
            session.stopped_by_user = True
            session.status = "stopping"
            session.updated_at = time.time()
        process = session.process
        try:
            if callable(getattr(process, "poll", None)) and process.poll() is None:
                if callable(getattr(process, "terminate", None)):
                    process.terminate()
                if callable(getattr(process, "wait", None)):
                    try:
                        process.wait(timeout=WATCH_STOP_TIMEOUT_SECONDS)
                    except TypeError:
                        process.wait()
                    except subprocess.TimeoutExpired:
                        if callable(getattr(process, "kill", None)):
                            process.kill()
                        process.wait(timeout=WATCH_STOP_TIMEOUT_SECONDS)
        finally:
            with session._lock:
                if callable(getattr(process, "poll", None)):
                    session.exit_code = process.poll()
                session.status = "stopped"
                session.updated_at = time.time()


resource_watch_manager = ResourceWatchManager()


def stop_all_resource_watches() -> dict[str, Any]:
    return resource_watch_manager.stop_all()

File: apps/desktop/src/main/backend/watch/watchManager.ts
import {
  spawn,
  type ChildProcessWithoutNullStreams,
} from "node:child_process";
import { randomUUID } from "node:crypto";
import { createInterface } from "node:readline";

import type { ResourceSnapshotCache } from "../cache/resourceSnapshotCache";
import {
  buildKubectlCommand,
  type KubectlCommand,
} from "../kubectl/command";
import {
  sanitizeKubectlText,
  truncateKubectlText,
} from "../kubectl/errors";
import type { SpawnProcess } from "../kubectl/runner";
import type { ResourceWatchEventHub } from "./eventHub";

const WATCH_TAIL_LINES = 20;
const WATCH_TAIL_LINE_CHARS = 1000;
const WATCH_STOP_TIMEOUT_MS = 3000;

export type WatchStatus = "running" | "stopping" | "stopped" | "failed";

interface WatchKey {
  clusterId: string;
  resource: string;
  namespace: string;
}

interface WatchSession {
  id: string;
  key: WatchKey;
  commandPreview: string;
  process: ChildProcessWithoutNullStreams;
  startedAt: number;
  updatedAt: number;
  status: WatchStatus;
  stdoutLines: number;
  stderrLines: number;
  cacheEvents: number;
  cacheInvalidations: number;
  exitCode: number | null;
  stoppedByUser: boolean;
  outputTail: string[];
  errorTail: string[];
  closePromise: Promise<void>;
  resolveClose: () => void;
}

export interface WatchView {
  id: string;
  clusterId: string;
  resource: string;
  namespace: string;
  status: WatchStatus;
  pid: number | null;
  startedAt: number;
  updatedAt: number;
  ageSeconds: number;
  stdoutLines: number;
  stderrLines: number;
  cacheEvents: number;
  cacheInvalidations: number;
  exitCode: number | null;
  stoppedByUser: boolean;
  commandPreview: string;
  outputTail: string[];
  errorTail: string[];
}

export interface WatchStartResult extends WatchView {
  alreadyRunning: boolean;
}

export interface WatchManagerStatus {
  enabled: true;
  mode: "cache-invalidation+websocket-events";
  running: number;
  total: number;
  watches: WatchView[];
  note: string;
}

export class WatchStartError extends Error {
  constructor(
    readonly code: "KUBECTL_NOT_FOUND" | "WATCH_START_FAILED",
    message: string,
    readonly rawStderr: string,
    readonly commandPreview: string,
  ) {
    super(message);
  }
}

interface ParsedWatchEvent {
  eventType: string;
  namespace: string;
  name: string;
}

function normalizedKey(key: WatchKey): string {
  return `${key.clusterId}\u0000${key.resource.toLowerCase()}\u0000${key.namespace}`;
}

function normalizeNamespace(namespace: string): string {
  const value = namespace.trim();
  return value || "all";
}

function tailPush(target: string[], line: string): void {
  const sanitized = sanitizeKubectlText(line).slice(0, WATCH_TAIL_LINE_CHARS);
  target.push(sanitized);
  if (target.length > WATCH_TAIL_LINES) {
    target.splice(0, target.length - WATCH_TAIL_LINES);
  }
}

function parseWatchEvent(line: string): ParsedWatchEvent | null {
  const text = line.trim();
  if (!text.startsWith("{")) return null;
  try {
    const value: unknown = JSON.parse(text);
    if (!value || typeof value !== "object" || Array.isArray(value)) return null;
    const record = value as Record<string, unknown>;
    const object = record.object;
    if (!object || typeof object !== "object" || Array.isArray(object)) return null;
    const metadataValue = (object as Record<string, unknown>).metadata;
    const metadata =
      metadataValue && typeof metadataValue === "object" && !Array.isArray(metadataValue)
        ? (metadataValue as Record<string, unknown>)
        : {};
    const namespace =
      typeof metadata.namespace === "string" && metadata.namespace.trim()
        ? metadata.namespace.trim()
        : "_cluster";
    const name = typeof metadata.name === "string" ? metadata.name : "";
    const eventType =
      typeof record.type === "string" && record.type.trim()
        ? record.type.trim()
        : "OBJECT";
    return { eventType, namespace, name };
  } catch {
    return null;
  }
}

function waitForSpawn(child: ChildProcessWithoutNullStreams): Promise<void> {
  return new Promise((resolve, reject) => {
    const onSpawn = () => {
      child.off("error", onError);
      resolve();
    };
    const onError = (error: Error) => {
      child.off("spawn", onSpawn);
      reject(error);
    };
    child.once("spawn", onSpawn);
    child.once("error", onError);
  });
}

function waitForClose(session: WatchSession, timeoutMs: number): Promise<boolean> {
  if (session.process.exitCode !== null || session.status === "stopped" || session.status === "failed") {
    return Promise.resolve(true);
  }
  return Promise.race([
    session.closePromise.then(() => true),
    new Promise<boolean>((resolve) => setTimeout(() => resolve(false), timeoutMs)),
  ]);
}

export class WatchManager {
  private readonly sessions = new Map<string, WatchSession>();
  private readonly runningByKey = new Map<string, string>();
  private closed = false;

  constructor(
    private readonly log: (message: string) => void,
    private readonly cache: ResourceSnapshotCache,
    private readonly eventHub: ResourceWatchEventHub,
    private readonly spawnProcess: SpawnProcess = spawn as SpawnProcess,
    private readonly now: () => number = Date.now,
    private readonly stopTimeoutMs = WATCH_STOP_TIMEOUT_MS,
  ) {}

  async start(
    command: KubectlCommand,
    resource: string,
    namespace = "all",
  ): Promise<WatchStartResult> {
    if (this.closed) {
      throw new WatchStartError(
        "WATCH_START_FAILED",
        "Watch manager is stopped",
        "",
        "",
      );
    }
    const key: WatchKey = {
      clusterId: command.clusterId,
      resource: resource.trim().toLowerCase(),
      namespace: normalizeNamespace(namespace),
    };
    const keyText = normalizedKey(key);
    const existingId = this.runningByKey.get(keyText);
    if (existingId) {
      const existing = this.sessions.get(existingId);
      if (existing && existing.status === "running") {
        return { ...this.view(existing), alreadyRunning: true };
      }
      this.runningByKey.delete(keyText);
    }

    const built = buildKubectlCommand(command);
    let child: ChildProcessWithoutNullStreams;
    try {
      child = this.spawnProcess(built.executable, built.args, {
        shell: false,
        windowsHide: true,
        env: built.environment,
      });
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error);
      const missing = (error as NodeJS.ErrnoException)?.code === "ENOENT";
      throw new WatchStartError(
        missing ? "KUBECTL_NOT_FOUND" : "WATCH_START_FAILED",
        missing ? `kubectl not found: ${command.kubectlPath}` : "kubectl watch could not be started",
        truncateKubectlText(sanitizeKubectlText(message)),
        built.preview,
      );
    }

    const startedAt = this.now() / 1000;
    let resolveClose!: () => void;
    const closePromise = new Promise<void>((resolve) => {
      resolveClose = resolve;
    });
    const session: WatchSession = {
      id: randomUUID(),
      key,
      commandPreview: built.preview,
      process: child,
      startedAt,
      updatedAt: startedAt,
      status: "running",
      stdoutLines: 0,
      stderrLines: 0,
      cacheEvents: 0,
      cacheInvalidations: 0,
      exitCode: null,
      stoppedByUser: false,
      outputTail: [],
      errorTail: [],
      closePromise,
      resolveClose,
    };

    const stdoutReader = createInterface({ input: child.stdout, crlfDelay: Infinity });
    const stderrReader = createInterface({ input: child.stderr, crlfDelay: Infinity });
    stdoutReader.on("line", (line) => this.handleStdoutLine(session, line));
    stderrReader.on("line", (line) => {
      session.stderrLines += 1;
      session.updatedAt = this.now() / 1000;
      tailPush(session.errorTail, line);
    });

    child.on("error", (error: NodeJS.ErrnoException) => {
      session.updatedAt = this.now() / 1000;
      tailPush(session.errorTail, error.message);
      if (session.status !== "stopping" && session.status !== "stopped") {
        session.status = "failed";
      }
      this.runningByKey.delete(keyText);
      resolveClose();
    });
    child.on("close", (code) => {
      session.exitCode = typeof code === "number" ? code : null;
      session.updatedAt = this.now() / 1000;
      if (session.status === "stopping" || session.stoppedByUser) {
        session.status = "stopped";
      } else if (code === 0) {
        session.status = "stopped";
      } else {
        session.status = "failed";
      }
      this.runningByKey.delete(keyText);
      stdoutReader.close();
      stderrReader.close();
      resolveClose();
      this.log(
        `node watch stopped id=${session.id} status=${session.status} exitCode=${String(session.exitCode)}`,
      );
    });
    child.stdin.on("error", () => {
      // stdin is intentionally closed for kubectl watch.
    });
    child.stdin.end();
    this.sessions.set(session.id, session);
    this.runningByKey.set(keyText, session.id);

    try {
      await waitForSpawn(child);
    } catch (error) {
      this.sessions.delete(session.id);
      this.runningByKey.delete(keyText);
      stdoutReader.close();
      stderrReader.close();
      const message = error instanceof Error ? error.message : String(error);
      const missing = (error as NodeJS.ErrnoException)?.code === "ENOENT";
      throw new WatchStartError(
        missing ? "KUBECTL_NOT_FOUND" : "WATCH_START_FAILED",
        missing ? `kubectl not found: ${command.kubectlPath}` : "kubectl watch could not be started",
        truncateKubectlText(sanitizeKubectlText(message)),
        built.preview,
      );
    }

    this.log(`node watch started id=${session.id} preview=${built.preview}`);
    return { ...this.view(session), alreadyRunning: false };
  }

  private handleStdoutLine(session: WatchSession, line: string): void {
    session.stdoutLines += 1;
    session.updatedAt = this.now() / 1000;
    tailPush(session.outputTail, line);
    const parsed = parseWatchEvent(line);
    if (!parsed) return;

    session.cacheEvents += 1;
    const cleared = this.cache.clearResource(
      session.key.clusterId,
      session.key.resource,
      parsed.namespace,
      "watch.event",
    );
    session.cacheInvalidations += cleared;
    this.eventHub.publish({
      type: "resource.changed",
      clusterId: session.key.clusterId,
      watchId: session.id,
      resource: session.key.resource,
      namespace: parsed.namespace,
      name: parsed.name,
      eventType: parsed.eventType,
      cacheInvalidations: cleared,
    });
  }

  status(): WatchManagerStatus {
    const watches = [...this.sessions.values()]
      .map((session) => this.view(session))
      .sort((a, b) => b.startedAt - a.startedAt);
    return {
      enabled: true,
      mode: "cache-invalidation+websocket-events",
      running: watches.filter((watch) => watch.status === "running").length,
      total: watches.length,
      watches,
      note:
        "Running watches parse kubectl watch events, invalidate matching Node resource snapshots, and publish WebSocket events. HTTP polling remains the fallback.",
    };
  }

  activeCount(): number {
    return [...this.sessions.values()].filter(
      (session) => session.status === "running" || session.status === "stopping",
    ).length;
  }

  async stop(id: string, stoppedByUser = true): Promise<{
    ok: boolean;
    found: boolean;
    id: string;
    watch?: WatchView;
  }> {
    const session = this.sessions.get(id);
    if (!session) return { ok: false, found: false, id };
    await this.stopSession(session, stoppedByUser);
    return { ok: true, found: true, id, watch: this.view(session) };
  }

  async stopCluster(clusterId: string): Promise<number> {
    const sessions = [...this.sessions.values()].filter(
      (session) =>
        session.key.clusterId === clusterId &&
        (session.status === "running" || session.status === "stopping"),
    );
    await Promise.all(sessions.map((session) => this.stopSession(session, false)));
    return sessions.length;
  }

  async stopAll(stoppedByUser = true): Promise<{
    ok: true;
    stopped: number;
    watches: WatchView[];
  }> {
    const sessions = [...this.sessions.values()].filter(
      (session) => session.status === "running" || session.status === "stopping",
    );
    await Promise.all(sessions.map((session) => this.stopSession(session, stoppedByUser)));
    return {
      ok: true,
      stopped: sessions.length,
      watches: sessions.map((session) => this.view(session)),
    };
  }

  private async stopSession(session: WatchSession, stoppedByUser: boolean): Promise<void> {
    if (session.status === "stopped" || session.status === "failed") return;
    session.stoppedByUser = session.stoppedByUser || stoppedByUser;
    session.status = "stopping";
    session.updatedAt = this.now() / 1000;
    try {
      if (!session.process.killed) session.process.kill();
    } catch (error) {
      tailPush(
        session.errorTail,
        error instanceof Error ? error.message : String(error),
      );
    }
    const closed = await waitForClose(session, this.stopTimeoutMs);
    if (!closed && session.process.exitCode === null) {
      try {
        session.process.kill("SIGKILL");
      } catch (error) {
        tailPush(
          session.errorTail,
          error instanceof Error ? error.message : String(error),
        );
      }
      await waitForClose(session, 1000);
    }
    if (session.status === "stopping") {
      session.status = "stopped";
      session.updatedAt = this.now() / 1000;
    }
    this.runningByKey.delete(normalizedKey(session.key));
  }

  async close(): Promise<void> {
    if (this.closed) return;
    this.closed = true;
    await this.stopAll(false);
  }

  private view(session: WatchSession): WatchView {
    const now = this.now() / 1000;
    return {
      id: session.id,
      clusterId: session.key.clusterId,
      resource: session.key.resource,
      namespace: session.key.namespace,
      status: session.status,
      pid: typeof session.process.pid === "number" ? session.process.pid : null,
      startedAt: session.startedAt,
      updatedAt: session.updatedAt,
      ageSeconds: Math.max(0, now - session.startedAt),
      stdoutLines: session.stdoutLines,
      stderrLines: session.stderrLines,
      cacheEvents: session.cacheEvents,
      cacheInvalidations: session.cacheInvalidations,
      exitCode: session.exitCode,
      stoppedByUser: session.stoppedByUser,
      commandPreview: session.commandPreview,
      outputTail: [...session.outputTail],
      errorTail: [...session.errorTail],
    };
  }
}
